SIMD: add a get_remaining_compute_units syscall

ckamm opened the pull request:
## Summary

Add a new syscall to access the number of compute units that are left to use in a transaction.

## Motivation

Some instructions can use a variable amount of compute. Examples include:

- Event processing instructions that can process as many events as are available.
- Orderbook matching instructions, which want to match a large taker order as  deeply into the orderbook as is necessary to fully fill it.
- Routing instructions, that may want to pick routes depending on remaining compute.

Currently these types of instructions often require the user to pass in a limit (like the maximum number of events to process, for the first example). That means the user needs to request the right compute budget for the transaction and guess the best possible limits for their instructions.

This often leads to users over-provisioning compute budget for a transaction. If the instruction could access the remaining budget, it could decide on its own whether another round of processing would fit or not, leading to better utilization of requested compute.

Example:
```
fn process_events() {
  while (true) {
    let event = peek_event();
    match event.type {
      CheapEvent => {
        if sol_remaining_compute_units() < 10_000 {
          break;
        }
        process_cheap(event);
      },
      ExpensiveEvent => {
        if sol_remaining_compute_units() < 40_000 {
          break;
        }
        process_expensive(event);
      }
    }
    pop_event();
  }
}
```

The example is inspired by an actual potential usecase in the Mango program. It shows how it would be preferable to "process events for all the 150k CU that remain" (so up to 15 cheap ones) instead of the current "process at most three events" (because 4 expensive ones would not fit).

7layermagik commented on proposals/0049-syscall-get-remaining-compute-units.md:
```diff
@@ -0,0 +1,104 @@
+(so up to 15 cheap ones) instead of the current "process at most three events"
+(because 4 expensive ones would not fit).
+
+## Alternatives Considered
```
It also seems important to figure out whether there will be any desire to use the compute budget request for figuring out block packing (i.e. bankless leader) one day so that there arent issues down the road.

mschneider replied:
i like this idea a lot, can we make this a follow up SIMD?

mschneider replied:
yeah seems to me also like something a good implementation could optimize around in jit stage. could this be the first 1 CU syscall?

mvines commented on proposals/0049-syscall-get-remaining-compute-units.md:
```diff
@@ -0,0 +1,104 @@
+## Detailed Design
+
+Add a syscall `sol_remaining_compute_units() -> u64` that returns the remaining
+compute budget for the current transaction.
```
Maybe also document the CU cost of calling this syscall to figure out the CUs remaining? (base syscall CU cost from the prototype impl, which seems plenty)

mvines commented on proposals/0049-syscall-get-remaining-compute-units.md:
```diff
@@ -0,0 +1,104 @@
+- Guess the effect of a CPI call by observing how much compute it used. But since
+  program and data bytes are public, that shouldn't create new capabilities.
+
+The compute budget would change from something invisible to programs to
+something accessible. Currently validators might be able to run transactions with
+half the requested budget only and then discard those that don't fit? Since the
+evaluation result of passing transactions currently doesn't depend on the compute
+budget? (not sure about this)
```
This feels like potential block optimization hack than a security consideration. I'd just remove this paragraph entirely, doesn't feel like it adds much. No strong pref either way though

mvines reviewed (approved):
I'm pretty supportive of this new facility. Don't see much downsides: API surface is small, implementation (in the Labs validator) is straight forward, allows programs better use what the user paid for

mvines commented:
@ckamm if you can get this PR green, I think we have sufficient consensus to merge it

ckamm commented:
@mvines I've incorporated the fixes and hopefully made the lints pass. (the one about expected headings had a weird error - I assume it didn't like the level-3 heading I had added, and have removed it).

What do you think about also covering a syscall to get the CU used by the current instruction in this SIMD? (see the "Alternatives Considered" section). I think for uses that involve composition of multiple instructions, it may be preferable to know how much compute the current instruction has already used up.

mvines commented:
> @mvines I've incorporated the fixes and hopefully made the lints pass. (the one about expected headings had a weird error - I assume it didn't like the level-3 heading I had added, and have removed it).

Looks like there's some more still! Since you're a first-time contributor here, somebody needs to mash a button to get CI to run. Just ping me again when you need a finger to do that.

> What do you think about also covering a syscall to get the CU used by the current instruction in this SIMD? (see the "Alternatives Considered" section). I think for uses that involve composition of multiple instructions, it may be preferable to know how much compute the current instruction has already used up.

The program could just call and store the value from `get_remaining_compute_units()` when it starts executing as a reference point right? No additional runtime support needed

ckamm commented:
The linter passes locally now (actually, I think it's buggy; it just requires at least one more section after the required ones).

> The program could just call and store the value from get_remaining_compute_units() when it starts executing as a reference point right? No additional runtime support needed

Doh, yeah. Agreed!

0xSol commented:
Per [current SIMD process](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0001-simd-process.md), This proposal reaches 6 months without activity. Marked as stagnant and to be closed. A new proposal can be opened if the proposal is closed and has a chance of reaching consensus.

lheeger-jump commented:
No. This SIMD needs some amendment before it is implementable. It is underspecified on specific details related to how CUs prior to a call to this function are to be computed.

willhickey commented:
Hey @t-nelson you were the approver on [the original PR](https://github.com/solana-labs/solana/pull/31640)

FD would rather not activate the associated feature gate. Do you have an opinion here?

t-nelson commented:
@0xSol does anyone actually still want this? afaik mango was the only proponent and they no longer exist. i never found the use cases that it enables particularly compelling, despite having okayed the implementation

t-nelson commented:
sry cc/ @Benhawkins18

t-nelson commented:
@Benhawkins18 mark as abandoned?

ckamm commented:
Sounds fine to me, I'm not continuing to push for this feature.
